cmd/solana_exporter/slots.go

```diff
@@ -35,88 +35,102 @@ type SlotWatcher struct {
+		TotalTransactionsMetric: prometheus.NewGauge(prometheus.GaugeOpts{
+			Name: "solana_confirmed_transactions_total",
+			Help: "Total number of transactions processed since genesis (max confirmation)",
+		}),
+		SlotHeightMetric: prometheus.NewGauge(prometheus.GaugeOpts{
+			Name: "solana_confirmed_slot_height",
+			Help: "Last confirmed slot height processed by watcher routine (max confirmation)",
+		}),
+		EpochNumberMetric: prometheus.NewGauge(prometheus.GaugeOpts{
+			Name: "solana_confirmed_epoch_number",
+			Help: "Current epoch (max confirmation)",
+		}),
+		EpochFirstSlotMetric: prometheus.NewGauge(prometheus.GaugeOpts{
+			Name: "solana_confirmed_epoch_first_slot",
+			Help: "Current epoch's first slot (max confirmation)",
+		}),
+		EpochLastSlotMetric: prometheus.NewGauge(prometheus.GaugeOpts{
+			Name: "solana_confirmed_epoch_last_slot",
+			Help: "Current epoch's last slot (max confirmation)",
+		}),
+		LeaderSlotsTotalMetric: prometheus.NewCounterVec(
+			prometheus.CounterOpts{
+				Name: "solana_leader_slots_total",
+				Help: "(DEPRECATED) Number of leader slots per leader, grouped by skip status",
+			},
+			[]string{SkipStatusLabel, NodekeyLabel},
+		),
+		LeaderSlotsByEpochMetric: prometheus.NewCounterVec(
+			prometheus.CounterOpts{
+				Name: "solana_leader_slots_by_epoch",
+				Help: "Number of leader slots per leader, grouped by skip status and epoch",
+			},
+			[]string{SkipStatusLabel, NodekeyLabel, EpochLabel},
+		),
+		InflationRewardsMetric: prometheus.NewGaugeVec(
+			prometheus.GaugeOpts{
+				Name: "solana_inflation_rewards",
+				Help: "Inflation reward earned per validator vote account, per epoch",
+			},
+			[]string{VotekeyLabel, EpochLabel},
+		),
+		FeeRewardsMetric: prometheus.NewCounterVec(
+			prometheus.CounterOpts{
+				Name: "solana_fee_rewards",
+				Help: "Transaction fee rewards earned per validator identity account, per epoch",
+			},
+			[]string{NodekeyLabel, EpochLabel},
+		),
 	}
-}
-
-func init() {
-	prometheus.MustRegister(totalTransactionsTotal)
-	prometheus.MustRegister(confirmedSlotHeight)
-	prometheus.MustRegister(currentEpochNumber)
-	prometheus.MustRegister(epochFirstSlot)
-	prometheus.MustRegister(epochLastSlot)
-	prometheus.MustRegister(leaderSlotsTotal)
-	prometheus.MustRegister(leaderSlotsByEpoch)
-	prometheus.MustRegister(inflationRewards)
-	prometheus.MustRegister(feeRewards)
+	// register:
+	for _, collector := range []prometheus.Collector{
+		watcher.TotalTransactionsMetric,
+		watcher.SlotHeightMetric,
+		watcher.EpochNumberMetric,
+		watcher.EpochFirstSlotMetric,
+		watcher.EpochLastSlotMetric,
+		watcher.LeaderSlotsTotalMetric,
+		watcher.LeaderSlotsByEpochMetric,
+		watcher.InflationRewardsMetric,
+		watcher.FeeRewardsMetric,
+	} {
+		if err := prometheus.Register(collector); err != nil {
+			var (
+				alreadyRegisteredErr *prometheus.AlreadyRegisteredError
+				duplicateErr         = strings.Contains(err.Error(), "duplicate metrics collector registration attempted")
```

Comment:
Why would this block ever be hit? Why would there ever be dupes when the list is hardcoded above and does not appear to have dupes?

Reply:
When you create multiple instances of `SlotWatcher`, such as when running tests.